Improve library documentation

Still34 opened the pull request:
Re-opening the PR since I accidentally deleted the old branch by renaming it.

### The PR will,
-	Fix minor grammatical errors.
-	Capitalize terms such as Commands, Modules and such, as the context is specific to the lib.
-	Wrap methods and properties in code blocks.
-	Fix broken samples and cross-reference links.
-	Clarify ambiguous terms or lines.


The current documentation suffers from several minor issues here and there; the PR aims to improve upon sections of the docs and add clarifications to ambiguous parts of the docs. 

Below are several issues that remain to be fixed or need to be clarified.
## Migration from 0.9


Sidebar linking doesn't work at all. Should we remove it altogether or fix it?


## Getting Started

### Installation

1. Should `UDPClient` still be referenced in installation?

---

### Your First Bot

1. > Connection logic is incomplete as of the current build. Events will soon be added to indicate when the client's state is ready for use (rewrite this section when possible).

Referring to the `Ready` event?

**Update**: Removed in 95935de.

2. > Handling a 'ping'...

Perhaps we should put a warning or info tooltip here telling the users that if they want to create a command, they should use `CommandService`?

**Update**: Added in b68528b.

## Command Service
---

1. >This article is out of date and has not been rewritten yet.
Information is not guaranteed to be accurate.

The docs doesn't necessarily seem "out of date" as the warning claims. The basics seem pretty relevant to the latest version of the lib. This warning turns off newcomers, as they might think that the official guide is unusable because of it being "outdated."

**Update**: Removed in f89aecb7.

---

2. >“To manually load a module, invoke [CommandService.AddModuleAsync], by passing in the generic type of your module and optionally, a dependency map.”

The latter part of the sentence seems unclear. Where should the user pass the dependency map to? 

---

3. >“First, you need to create an @System.IServiceProvider You may create your own IServiceProvider if you wish.” 

Any mention of @System.IServiceProvider are currently broken on the docs.


---

4. >“Submodules are Modules that reside within another one. Typically, submodules are used to create nested groups (although not required to create nested groups).”

Clarification on the part after "although?"


---

5. >“Finally, pass the map into the `LoadAssembly` method. Your modules will automatically be loaded with this dependency map.”

Where is this `LoadAssembly` method?


---

6. >Preconditions can be applied to Modules, Groups, or Commands.

The docs should mention `ParameterPreconditionAttribute`'s existence.

**Update**: Resolved in 4449a05.

Commit:
Improve the Command Service documentation

The following changes have been added to this PR:
•	Fix minor grammatical errors.
•	Capitalize terms such as Commands, Modules and such, as the context is specific to the lib.
•	Wrap methods and properties in code blocks.

The docs page currently has several issues that remains to be fixed.

1. 
```md
>[!WARNING]
>This article is out of date and has not been rewritten yet.
Information is not guaranteed to be accurate.
```

The docs doesn't necessarily seem "out of date" as the warning claims. The basics seem pretty relevant to the latest version of the lib.

2.
>“To manually load a module, invoke [CommandService.AddModuleAsync], by passing in the generic type of your module and optionally, a dependency map.”

The latter part of the sentence seems off. Where should the user pass the dependency map to? It seems to suggest that `AddModuleAsync` has an argument to pass the dependency to. If it is referring to `AddModuleAsync(Type type)`, then I feel like it should be clarified here - or perhaps change the wording of the sentence.

3.
>“First, you need to create an @System.IServiceProvider You may create your own IServiceProvider if you wish.” 

Any mention of @System.IServiceProvider is currently broken on the docs.

4.
>“Submodules are Modules that reside within another one. Typically, submodules are used to create nested groups (although not required to create nested groups).”

Clarification on the part after "although?"

5.
>“Finally, pass the map into the LoadAssembly method. Your modules will automatically be loaded with this dependency map.”

Where is this `LoadAssembly` method?

6.
```md
>[!NOTE]
>Preconditions can be applied to Modules, Groups, or Commands.
```

The docs should mention `ParameterPreconditionAttribute`'s existence.

Commit:
Update line breaks to comply with docs standard

Commit:
Change "you should..." to "instead, ..."

Commit:
Trim trailing spaces

Commit:
Change "inherits" to "inherit"

Commit:
Fix Context warning note and add ReplyAsync xref

Commit:
Fix broken xrefs

Commit:
Fix [Command Service] xref

Commit:
Fix consistency between TypeReaders and Preconditions returns

Commit:
Add missing semi-colons in ServiceProvider sample

Commit:
Change CommandContext to SocketCommandContext & change variable naming

Commit:
Cleanup TypeReader section

Commit:
Wrap [DontInject] in code block

Commit:
Fix commands docs linking in intro

Commit:
Improve Getting Started - Installation

- Fix character misalignment to comply with docs standard.
- Fix image numbering issues by moving the tooltips above some of the steps.
- Add codeblocks to search terms like `Discord.Net`.
- Remove broken `addons` reference.
- Specify `.NET 4.6.1` as `.NET Framework 4.6.1`.
- Minor cross-reference cleanup.

Commit:
Fix Getting Started - Intro

- Minor grammartical fixes.
- Wrap mentions of the methods, properties, and events in code block.
- Replace `Discord.Net` to `Discord.NET`.
- Fix steps numbering under `Creating a Discord Bot` and `Adding your bot to a server`.
- Change `Task-based Asynchronous Pattern ([TAP])` linking to mark the entire term instead.
- Change code block of `Pong!` to quotation mark instead.

Commit:
Fix cross references in Sending Voice

Still34 commented:
```diff
- >Preconditions can be applied to Modules, Groups, or Commands.
+ >There are two types of Preconditions. [PreconditionAttribute] can be applied to Modules, Groups, or Commands; [ParameterPreconditionAttribute] can be applied to Parameters.

- To write your own precondition, create a new class that inherits from @Discord.Commands.PreconditionAttribute.
+ To write your own Precondition, create a new class that inherits from either [PreconditionAttribute] or [ParameterPreconditionAttribute] depending on your use.

- In order for your precondition to function, you will need to override the [CheckPermissions] method.
+ In order for your Precondition to function, you will need to override the [CheckPermissions] method.
```

I propose this change to fix mention the use of `ParameterPreconditionAttribute`.

Commit:
Mention parameter precondition attribute

Commit:
Change `Discord.NET` to `Discord.Net` for consistency

Commit:
Wrap project names in code blocks & minor fixes in Terminology

Commit:
Change `add-ons` to `addons` for consistency

Commit:
Fix cross references in Logging

Commit:
Fix minor grammatical issues in "Working with Events"

Commit:
Missed a tilda

Commit:
Remove out-of-date warning in Commands

Commit:
Minor grammatical fixes for Entities

Commit:
Fix broken xref in Logging

Chris-Johnston commented on Discord.Net.sln:
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
-VisualStudioVersion = 15.0.26228.4
+VisualStudioVersion = 15.0.26730.12
```
I don't know if changes to .sln file will cause merge conflicts or not. Might be easier to exclude this file?

Still34 replied:
The person who'd merge the PR will just have to ignore this during merge

Chris-Johnston commented on docs/guides/commands/samples/dependency_map_setup.cs:
```diff
@@ -9,9 +9,9 @@ public class Commands
-        _serviceCollection.AddSingleton(new DatabaseService())
+        _serviceCollection.AddSingleton(client);
+        _serviceCollection.AddSingleton(new NotificationService());
+        _serviceCollection.AddSingleton(new DatabaseService());
```
While this file is meant to be a short example, maybe add the declarations of `_serviceCollection` and `_commands`?

AntiTcb replied:
Alongside this, I'd take out the class declaration and namespace imports, they don't really add anything to the example.

Still34 replied:
I agree. I'll modify it accordingly shortly.

Chris-Johnston commented on docs/guides/concepts/events.md:
```diff
@@ -4,27 +4,27 @@ title: Working with Events
 
-All events are designed to be thread-safe, in that events are executed
-synchronously off the gateway task, in the same context as the gateway
+All events are designed to be thread-safe in that events are executed
```
"All events are designed to be thread-safe; events are executed synchronously off the gateway task in the same context as the gateway task."
Original sentence kinda keeps going imo

Chris-Johnston commented on docs/guides/concepts/events.md:
```diff
@@ -33,7 +33,7 @@ offloaded to another task.
 from Discord's gateway in the same context of an event. Since the
 gateway will wait on all invoked event handlers to finish before
-processing any additional data from the gateway, this will create
+processing any additional data from the gateway; this will create
```
"Since the gateway will wait on all invoked event handlers to finish before processing any additional data from the gateway" is a dependent clause, I'd leave the comma.

Still34 replied:
Yup, that was a mistake on my part. Will fix.

Chris-Johnston commented on docs/guides/concepts/events.md:
```diff
@@ -76,7 +76,7 @@ object.
 ### Tips
 
-Many events relating to a Message entity, e.g. `MessageUpdated`
+Many events relating to a Message entity (e.g. `MessageUpdated`)
```
...to a Message entity (e.g. `MessageUpdated` and `ReactionAdded`) rely on...

Chris-Johnston commented on docs/guides/concepts/logging.md:
```diff
@@ -14,12 +14,14 @@ section.
 
 To receive log events, simply hook the discord client's log method
-to a Task with a single parameter of type [LogMessage]
+to a `Task` with a single parameter of type [LogMessage].
```
of type `[LogMessage]`.

~~is [LogMessage] a link to the docs? Maybe for consistency, I would change it to `LogMessage`~~ ah, just saw

Still34 replied:
Most bracket-contained words are xrefs.

AntiTcb commented on docs/guides/commands/samples/command_handler.cs:
```diff
@@ -8,36 +8,36 @@
+    private DiscordSocketClient _client;
+    private IServiceProvider _services;
 
     static void Main(string[] args) => new Program().Start().GetAwaiter().GetResult();
```
`static async Task Main(string[] args) => await new Program().StartAsync()`

Let's break out the C#7.1 features.

AntiTcb commented on docs/guides/commands/samples/command_handler.cs:
```diff
@@ -8,36 +8,36 @@
 
     static void Main(string[] args) => new Program().Start().GetAwaiter().GetResult();
 
     public async Task Start()
```
StartAsync*

Chris-Johnston commented on docs/guides/getting_started/intro.md:
```diff
@@ -99,33 +105,37 @@ the Console.
-**do __NOT__ share this token with anyone!** It may behoove you to
-store this token in an external file if you plan on distributing the
-source code for your bot.
+**do __NOT__ share this token with anyone else!** It may behoove you 
```
I think if we are going to stress that this is something really important, we should use straightforward wording. Swap "behoove" with "It is responsible to store this token in...", or other wording.

Still34 replied:
Won't fix. Main contribs preferred the word "behoove" to stay.

Chris-Johnston commented on docs/guides/getting_started/intro.md:
```diff
@@ -136,69 +146,68 @@ handler.
-soon be added to indicate when the client's state is ready for use;
-(rewrite this section when possible)
+soon be added to indicate when the client's state is ready for use
+(rewrite this section when possible).
```
outdated?
**EDIT:** I see you already mentioned this. Oops.
> Referring to the Ready event?

AntiTcb commented on docs/guides/commands/samples/command_handler.cs:
```diff
@@ -8,36 +8,36 @@
+        _client = new DiscordSocketClient();
+        _commands = new CommandService();
 
         string token = "bot token here";
```
This probably should be something other than the hard-coded method. Even something as simple as a `File.ReadAllText("config.txt")`, where `config.txt` is literally just the token in a text file would be good enough.

Chris-Johnston commented on docs/guides/getting_started/intro.md:
```diff
@@ -136,69 +146,68 @@ handler.
 [!code-csharp[Create client](samples/intro/client.cs)]
 
-At this point, feel free to start your program and see your bot come
+At this point, feel free to start your program and see your bot comes
```
The original wording was correct.
As an alternative, maybe try "If all goes well, upon starting your program your bot should appear online."

Still34 replied:
How so? The verb should be referring to `your bot`, no?

AntiTcb replied:
Future tense, you'd use "come." As in, "your bot will come online," as opposed to "your bot will comes online."

Still34 replied:
Ah, alright. I originally interpreted the sentence as "see if your bot comes online." Will fix.

Chris-Johnston commented on docs/guides/getting_started/intro.md:
```diff
@@ -136,69 +146,68 @@ handler.
-until after the application is exited. To do this, we can await an
-infinite delay, or any other blocking method, such as reading from
+until after the application has exited. To do this, we can await an
+infinite delay or any other blocking method, such as reading from
```
I think the previous commas were correct.

Chris-Johnston commented on docs/guides/getting_started/intro.md:
```diff
@@ -136,69 +146,68 @@ handler.
-MessageReceived event - it must be a method (`Func`) that returns the
-type `Task`, and takes a single parameter, a [SocketMessage]. Also,
+`MessageReceived` event - it must be a method (`Func`) that returns 
+the type `Task` and takes a single parameter, a [SocketMessage]. Also,
```
`[SocketMessage]` ?

Still34 replied:
What about it?

Chris-Johnston replied:
My mistake. I thought you could wrap links in accent marks to have a link inside a code block, but it can't be done. Please ignore.

AntiTcb commented on docs/guides/commands/samples/command_handler.cs:
```diff
@@ -8,36 +8,36 @@
 
-        services = new ServiceCollection()
-                .BuildServiceProvider();
+        _services = new ServiceCollection()
```
At the very least, this could inject `_client`, and `_commands` into the container.

AntiTcb commented on docs/guides/commands/samples/command_handler.cs:
```diff
@@ -8,36 +8,36 @@
+        _services = new ServiceCollection()
+            .BuildServiceProvider();
 
         await InstallCommands();
```
InstallCommandsAsync*

AntiTcb commented on docs/guides/commands/samples/command_handler.cs:
```diff
@@ -8,36 +8,36 @@
         await Task.Delay(-1);
     }
 
     public async Task InstallCommands()
```
InstallCommandsAsync*

AntiTcb commented on docs/guides/voice/sending-voice.md:
```diff
@@ -17,7 +17,7 @@ when developing on .NET Core, this is where you execute `dotnet run`
 
 For Windows Users, precompiled binaries are available for your 
-convienence [here](https://discord.foxbot.me/binaries/)
+convienence [here](https://discord.foxbot.me/binaries/).
```
Dead link, should be removed/replaced with Emzi's link.

Still34 replied:
Agree. Will have to get @foxbot's input first, though.

Commit:
Adjust service collection sample

...according to https://github.com/RogueException/Discord.Net/pull/826/files/f89aecb7bfa8fe82a541d3bba44d0e37c125d235#r141530227

Commit:
Update Command Handler sample

- Update Main for C# 7.1.
- Inject CommandService and DiscordSocketClient into the service collection.
- Add Async suffix to asynchronous methods.

Commit:
Minor grammatical fixes in Events

Commit:
Revert 2 incorrect grammar corrections

Still34 commented:
Should we break out async Main across all sample codes? @AntiTcb suggested that one of the samples should use async Main like `static Task Main(string[] args) => new Program().StartAsync();`. At the moment there are 8 other files that still use `static void Main`.

AntiTcb commented:
I'd prefer it as it means exceptions will be displayed normally instead of just being bubbled up to `GetAwaiter().GetResult()` and needing to explain that and how to enable breaking on all CLR exceptions.

Fair point though, C#7.1 features are still off by default in VS projects currently.

Joe4evr commented:
> I propose this change to mention the use of `ParameterPreconditionAttribute`.

I would say to not wrap the class names in brackets when also adding the `Attribute` part of it.

Still34 commented:
Oops, that's definitely an oversight on my part. Will fix. @Joe4evr

Commit:
Revert async Main sample

Commit:
Add hardcode token notice in sample

Commit:
Fix missing method for Command Handler

Commit:
Modify module samples to use SocketCommandContext instead

Commit:
Emphasize CommandContext and SocketCommandContext

Commit:
Fix formatting for module sample

Commit:
Add SocketCommandContext for Groups sample

Commit:
Remove comma

Commit:
Fix DepMap sample formatting

Commit:
Replace [DontInject] with DontInjectAttribute with cross reference

Commit:
Remove connection logic note

There is no reason that this note should still be here since Ready event exists.

Commit:
Add a new warning message informing the users the existence of CommandService

Commit:
Make command handler private

Still34 commented:
There are still several issues that remain to be fixed with the docs that require some discussion. I wasn't expecting the PR to be merged already.

khionu commented:
For future reference, you should denote that in the PR, to prevent this.
